lib/python/shape.cpp

```diff
@@ -51,9 +51,19 @@ template <class T> void bind_fold(pybind11::module &mod) {
 template <class T> void bind_flatten(pybind11::module &mod) {
   mod.def(
       "flatten",
-      [](const T &self, const std::vector<std::string> &dims,
+      [](const T &self, const std::optional<std::vector<std::string>> &dims,
          const std::string &to) {
-        return flatten(self, to_dim_type(dims), Dim{to});
+        if (dims.has_value())
+          return flatten(self, to_dim_type(*dims), Dim{to});
+        // If no dims are given then we flatten all dims. For variables we just
+        // provide a list of all labels. DataArrays are different, as the
+        // behavior in the degenerate case of a 0-D 'self' must distinguish
+        // between flattening "zero dims" and "all dims". The latter is
+        // specified using std::nullopt.
```

Comment:
Should we describe the behaviour better (what happens when no dims are supplied, what happens when empty dims are supplied) in the python docstring as this is what the users will see?

Reply:
Done, and also added another test, highlighting a case that was previously not working as it should (raising an exception).